L1Trigger/L1TTrackMatch/plugins/L1TrackTripletEmulatorProducer.cc

```diff
@@ -0,0 +1,452 @@
+      pair1_massMax_(iConfig.getParameter<double>("pair1_massMax")),
+      pair2_massMin_(iConfig.getParameter<double>("pair2_massMin")),
+      pair2_massMax_(iConfig.getParameter<double>("pair2_massMax")),
+      pair1_dzMin_(iConfig.getParameter<double>("pair1_dzMin")),
+      pair1_dzMax_(iConfig.getParameter<double>("pair1_dzMax")),
+      pair2_dzMin_(iConfig.getParameter<double>("pair2_dzMin")),
+      pair2_dzMax_(iConfig.getParameter<double>("pair2_dzMax")),
+      use_float_track_precision_(iConfig.getParameter<bool>("float_precision")),
+      trackToken_(consumes<L1TTTrackRefCollectionType>(iConfig.getParameter<InputTag>("L1TrackInputTag"))),
+      PVtxToken_(consumes<l1t::VertexWordCollection>(iConfig.getParameter<InputTag>("L1PVertexInputTag"))) {
+  if (displaced_) {
+    produces<l1t::TkTripletCollection>("L1TrackTripletExtended");
+    produces<l1t::TkTripletWordCollection>("L1TrackTripletWordExtended");
+  } else {
+    produces<l1t::TkTripletCollection>("L1TrackTriplet");
+    produces<l1t::TkTripletWordCollection>("L1TrackTripletWord");
+  }
+}
+
+void L1TrackTripletEmulatorProducer::produce(Event &iEvent, const EventSetup &iSetup) {
+  unique_ptr<l1t::TkTripletCollection> L1TrackTripletContainer(new l1t::TkTripletCollection);
+  unique_ptr<l1t::TkTripletWordCollection> L1TrackTripletWordContainer(new l1t::TkTripletWordCollection);
+
+  // Read inputs
+  edm::Handle<L1TTTrackRefCollectionType> TTTrackHandle;
+  iEvent.getByToken(trackToken_, TTTrackHandle);
+
+  edm::Handle<l1t::VertexWordCollection> PVtx;
+  iEvent.getByToken(PVtxToken_, PVtx);
+  double PVz = (PVtx->at(0)).z0();
+
+  std::string OutputDigisName = "L1TrackTriplet";
+  std::string OutputWordName = "L1TrackTripletWord";
+
+  if (displaced_)
+    OutputDigisName += "Extended";
+
+  if (TTTrackHandle->size() < 3) {
+    iEvent.put(std::move(L1TrackTripletContainer), OutputDigisName);
+    iEvent.put(std::move(L1TrackTripletWordContainer), OutputWordName);
+    return;
+  }
+
+  // Finding the 3 highest pT tracks
+  L1track trk1{-99, -99, -99, -99, -99, -99, -99, 0};
+  L1track trk2{-99, -99, -99, -99, -99, -99, -99, 0};
+  L1track trk3{-99, -99, -99, -99, -99, -99, -99, 0};
+
+  for (auto current_track : *TTTrackHandle) {
+    double current_track_pt = 0;
+    if (use_float_track_precision_)
+      current_track_pt = current_track->momentum().perp();
+    else
+      current_track_pt = FloatPtFromBits(*current_track);
+    if (current_track_pt < trk1.Pt)
+      continue;
+    trk3 = trk2;
+    trk2 = trk1;
+    if (use_float_track_precision_) {
+      trk1.Pt = current_track->momentum().perp();
+      trk1.Eta = current_track->eta();
+      trk1.Phi = current_track->phi();
+      trk1.Charge = (int)(current_track->rInv() / fabs(current_track->rInv()));
+      trk1.MVA = current_track->trkMVA1();
+      trk1.Nstubs = current_track->getStubRefs().size();
+      trk1.Z0 = current_track->z0();
+      trk1.Index = &(*current_track) - &(*TTTrackHandle->at(0));
```

Comment:
Sure, I will. The trk1 is a simple struct to store attributes of the TTTracks. This is because I want to have a single code both for simulation and emulation. The index is defined in this line [*].  What this line does is to retrieve the index of the track in range-based loops. I can do the same thing with an increasing counter but since this part is only meant to run on emulation and not firmware, I used this more elegant line.


[*] https://github.com/cms-l1t-offline/cmssw/blob/a56001e9857d26c2d26e51a67c878abeeff2c7ad/L1Trigger/L1TTrackMatch/plugins/L1TrackTripletEmulatorProducer.cc#L107

Reply:
Let me make sure I have this correct. You get a `L1TTTrackRefCollectionType` from the dereferenced `TTrackHandle`, you then dereference that with `*`, but then get a pointer/memory address with `&`? And do memory arithmetic with the dereferenced object's address to derive the index? I think I see the logic here, but it seems a bit roundabout to me, and I'm a bit skittish about the correctness of doing anything with bare memory addresses if it can be avoided. 

Ultimately it's up to you here, If you've tested this works then I won't be picky, but I think the increasing counter be a bit more future-proof.

Reply:
I did test it and works fine, but I changed it to a counter. I dont have strong opinion anyway